File: modules/time-tracking/src/main/java/com/boom/timetracking/SessionBuilder.java
package com.boom.timetracking;

import java.util.ArrayList;
import java.util.Collections;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Stateful builder that consumes {@link UsageEvent}s and emits completed {@link Session} objects.
 */
public final class SessionBuilder {

    public static final long DEFAULT_INACTIVITY_THRESHOLD = 15_000L;
    public static final long DEFAULT_STRAY_BACKGROUND_WINDOW = 5_000L;

    public static final class Stats {
        private final long emittedSessions;
        private final long strayBackgroundSessions;
        private final long autoRepairedSessions;

        public Stats(long emittedSessions, long strayBackgroundSessions, long autoRepairedSessions) {
            this.emittedSessions = emittedSessions;
            this.strayBackgroundSessions = strayBackgroundSessions;
            this.autoRepairedSessions = autoRepairedSessions;
        }

        public long getEmittedSessions() {
            return emittedSessions;
        }

        public long getStrayBackgroundSessions() {
            return strayBackgroundSessions;
        }

        public long getAutoRepairedSessions() {
            return autoRepairedSessions;
        }
    }

    private static final class MutableState {
        private final String packageName;
        private long start;
        private long lastEvent;
        private int eventCount;
        private final EnumSet<UsageEvent.Source> sources;
        private final LinkedHashMap<String, String> attributes;

        private MutableState(String packageName, long start, UsageEvent.Source source, Map<String, String> metadata) {
            this.packageName = packageName;
            this.start = start;
            this.lastEvent = start;
            this.eventCount = 1;
            this.sources = EnumSet.of(source);
            this.attributes = new LinkedHashMap<>();
            if (metadata != null) {
                this.attributes.putAll(metadata);
            }
        }

        private void update(UsageEvent event) {
            this.lastEvent = event.getTimestamp();
            this.eventCount += 1;
            this.sources.add(event.getSource());
            if (!event.getMetadata().isEmpty()) {
                this.attributes.putAll(event.getMetadata());
            }
        }

        private Session toSession(long endTime, Session.RepairStatus repairStatus) {
            return new Session(
                packageName,
                start,
                endTime,
                eventCount,
                sources,
                determineConfidence(sources),
                repairStatus,
                attributes
            );
        }
    }

    private final long inactivityThresholdMillis;
    private final long strayBackgroundWindowMillis;
    private long emitted = 0L;
    private long strayBackground = 0L;
    private long autoRepaired = 0L;
    private final Map<String, MutableState> activeStates = new HashMap<>();

    public SessionBuilder() {
        this(DEFAULT_INACTIVITY_THRESHOLD, DEFAULT_STRAY_BACKGROUND_WINDOW);
    }

    public SessionBuilder(long inactivityThresholdMillis, long strayBackgroundWindowMillis) {
        this.inactivityThresholdMillis = inactivityThresholdMillis;
        this.strayBackgroundWindowMillis = strayBackgroundWindowMillis;
    }

    public Stats stats() {
        return new Stats(emitted, strayBackground, autoRepaired);
    }

    public void reset() {
        emitted = 0;
        strayBackground = 0;
        autoRepaired = 0;
        activeStates.clear();
    }

    public List<Session> process(UsageEvent event) {
        switch (event.getEventType()) {
            case FOREGROUND:
                return handleForeground(event);
            case BACKGROUND:
                return handleBackground(event);
            default:
                return Collections.emptyList();
        }
    }

    public List<Session> flush(long now) {
        List<Session> completed = new ArrayList<>();
        List<String> toRemove = new ArrayList<>();
        for (Map.Entry<String, MutableState> entry : activeStates.entrySet()) {
            MutableState state = entry.getValue();
            if (now - state.lastEvent >= inactivityThresholdMillis) {
                completed.add(state.toSession(state.lastEvent, Session.RepairStatus.AUTO_REPAIRED));
                autoRepaired += 1;
                emitted += 1;
                toRemove.add(entry.getKey());
            }
        }
        for (String key : toRemove) {
            activeStates.remove(key);
        }
        return completed;
    }

    private List<Session> handleForeground(UsageEvent event) {
        MutableState state = activeStates.get(event.getPackageName());
        if (state == null) {
            activeStates.put(event.getPackageName(), new MutableState(
                event.getPackageName(), event.getTimestamp(), event.getSource(), event.getMetadata()
            ));
            return Collections.emptyList();
        }
        long gap = event.getTimestamp() - state.lastEvent;
        if (gap > inactivityThresholdMillis) {
            Session closed = state.toSession(state.lastEvent, Session.RepairStatus.AUTO_REPAIRED);
            autoRepaired += 1;
            emitted += 1;
            activeStates.put(event.getPackageName(), new MutableState(
                event.getPackageName(), event.getTimestamp(), event.getSource(), event.getMetadata()
            ));
            return Collections.singletonList(closed);
        }
        state.update(event);
        return Collections.emptyList();
    }

    private List<Session> handleBackground(UsageEvent event) {
        MutableState state = activeStates.get(event.getPackageName());
        if (state == null) {
            Session repaired = strayBackgroundSession(event);
            emitted += 1;
            strayBackground += 1;
            autoRepaired += 1;
            return Collections.singletonList(repaired);
        }
        long gap = event.getTimestamp() - state.lastEvent;
        if (gap > inactivityThresholdMillis) {
            List<Session> sessions = new ArrayList<>();
            Session closed = state.toSession(state.lastEvent, Session.RepairStatus.AUTO_REPAIRED);
            autoRepaired += 1;
            emitted += 1;
            sessions.add(closed);
            activeStates.remove(event.getPackageName());
            if (gap <= strayBackgroundWindowMillis) {
                Session repaired = strayBackgroundSession(event);
                sessions.add(repaired);
                strayBackground += 1;
                autoRepaired += 1;
                emitted += 1;
            }
            return sessions;
        }
        state.update(event);
        Session session = state.toSession(event.getTimestamp(), Session.RepairStatus.NONE);
        emitted += 1;
        activeStates.remove(event.getPackageName());
        return Collections.singletonList(session);
    }

    private Session strayBackgroundSession(UsageEvent event) {
        return new Session(
            event.getPackageName(),
            event.getTimestamp(),
            event.getTimestamp(),
            1,
            Collections.singleton(event.getSource()),
            Session.Confidence.LOW,
            Session.RepairStatus.AUTO_REPAIRED,
            event.getMetadata()
        );
    }

    private static Session.Confidence determineConfidence(EnumSet<UsageEvent.Source> sources) {
        boolean hasUsageStats = sources.contains(UsageEvent.Source.USAGE_STATS);
        boolean hasAccessibility = sources.contains(UsageEvent.Source.ACCESSIBILITY);
        if (hasUsageStats && hasAccessibility) {
            return Session.Confidence.HIGH;
        }
        if (hasUsageStats) {
            return Session.Confidence.HIGH;
        }
        if (hasAccessibility) {
            return Session.Confidence.MEDIUM;
        }
        return Session.Confidence.LOW;
    }
}

File: modules/time-tracking/src/main/java/com/boom/timetracking/SessionBuilder.kt
package com.boom.timetracking

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch

/**
 * Consumes [UsageEvent] stream, builds [Session] objects and persists them via [SessionWriter].
 */
class SessionBuilder(
    private val scope: CoroutineScope,
    private val sessionWriter: SessionWriter,
    private val inactivityThresholdMillis: Long = DEFAULT_INACTIVITY_THRESHOLD
) : UsageEventCollector.UsageEventSink {

    interface SessionWriter {
        suspend fun upsert(session: Session)
    }

    private val eventsFlow = MutableSharedFlow<UsageEvent>(
        replay = 0,
        extraBufferCapacity = 256,
        onBufferOverflow = BufferOverflow.DROP_OLDEST
    )

    private var buildJob: Job? = null
    private val openSessions = mutableMapOf<String, MutableSessionAccumulator>()

    fun start() {
        buildJob?.cancel()
        buildJob = scope.launch(Dispatchers.Default) {
            eventsFlow.collect { event ->
                handleEvent(event)
            }
        }
    }

    override suspend fun emit(event: UsageEvent) {
        eventsFlow.emit(event)
    }

    fun stop() {
        buildJob?.cancel()
        buildJob = null
        openSessions.clear()
    }

    suspend fun flushActiveSessions(now: Long) {
        openSessions.entries.toList().forEach { (pkg, accumulator) ->
            if (accumulator.eventCount == 0) return@forEach
            val session = Session(
                packageName = pkg,
                startTime = accumulator.startTime,
                endTime = now,
                eventCount = accumulator.eventCount,
                confidence = Session.Confidence.LOW,
                sourceMask = SourceMask.fromSources(accumulator.sources)
            )
            sessionWriter.upsert(session)
        }
        openSessions.clear()
    }

    private suspend fun handleEvent(event: UsageEvent) {
        val accumulator = openSessions.getOrPut(event.packageName) { MutableSessionAccumulator() }
        when (event.eventType) {
            UsageEvent.EventType.FOREGROUND -> accumulator.onForeground(event)
            UsageEvent.EventType.BACKGROUND -> {
                val session = accumulator.onBackground(event, inactivityThresholdMillis)
                if (session != null) {
                    sessionWriter.upsert(session)
                    openSessions.remove(event.packageName)
                } else if (accumulator.eventCount == 0) {
                    openSessions.remove(event.packageName)
                }
            }
        }
    }

    private inner class MutableSessionAccumulator {
        var startTime: Long = 0
            private set
        var lastEventTime: Long = 0
            private set
        var eventCount: Int = 0
            private set
        val sources = mutableSetOf<UsageEvent.Source>()

        fun onForeground(event: UsageEvent) {
            if (eventCount == 0) {
                startTime = event.timestamp
            }
            lastEventTime = event.timestamp
            eventCount += 1
            sources += event.source
        }

        suspend fun onBackground(
            event: UsageEvent,
            inactivityThresholdMillis: Long
        ): Session? {
            if (eventCount == 0) {
                // Received background event without foreground; ignore but avoid leak
                return null
            }
            val gap = event.timestamp - lastEventTime
            return if (gap <= inactivityThresholdMillis) {
                eventCount += 1
                lastEventTime = event.timestamp
                sources += event.source
                Session(
                    packageName = event.packageName,
                    startTime = startTime,
                    endTime = event.timestamp,
                    eventCount = eventCount,
                    confidence = when {
                        UsageEvent.Source.ACCESSIBILITY in sources && UsageEvent.Source.USAGE_STATS in sources -> Session.Confidence.HIGH
                        UsageEvent.Source.USAGE_STATS in sources -> Session.Confidence.HIGH
                        else -> Session.Confidence.MEDIUM
                    },
                    sourceMask = SourceMask.fromSources(sources)
                )
            } else {
                // Gap too large, treat as orphan background -> flush existing session
                val session = Session(
                    packageName = event.packageName,
                    startTime = startTime,
                    endTime = lastEventTime,
                    eventCount = eventCount,
                    confidence = Session.Confidence.MEDIUM,
                    sourceMask = SourceMask.fromSources(sources)
                )
                startTime = 0
                lastEventTime = 0
                eventCount = 0
                sources.clear()
                sessionWriter.upsert(session)
                null
            }
        }
    }

    companion object {
        const val DEFAULT_INACTIVITY_THRESHOLD = 20_000L
    }
}
